Error: "Cannot read property 'loc' of undefined"

# Minimum operation for the error

```bash
mkdir -p /tmp/err_sample
cd  /tmp/err_sample
npm install textlint textlint-plugin-latex2e textlint-rule-no-nfd
echo '{"plugins":["latex2e"],"rules":{"no-nfd":true}}' > .textlintrc
echo -e 'A%B\nC' > sample.tex
npx textlint ./sample.tex
```

No error with the following setting.


```bash
echo '{"rules":{"no-nfd":true}}' > .textlintrc
npx textlint ./sample.tex
```
Here is [package-lock.json](https://github.com/textlint/textlint-plugin-latex2e/files/5746541/package-lock.json.txt).

# Error message

```txt
✖ Error
Cannot read property 'loc' of undefined

✖ Stack trace
TypeError: Cannot read property 'loc' of undefined
    at validate (/private/tmp/err_sample/node_modules/@textlint/kernel/lib/context/TextlintSourceCodeImpl.js:37:14)
    at new TextlintSourceCodeImpl (/private/tmp/err_sample/node_modules/@textlint/kernel/lib/context/TextlintSourceCodeImpl.js:56:9)
    at TextlintKernel._parallelProcess (/private/tmp/err_sample/node_modules/@textlint/kernel/lib/textlint-kernel.js:153:26)
    at /private/tmp/err_sample/node_modules/@textlint/kernel/lib/textlint-kernel.js:100:26
npm ERR! code 1
npm ERR! path /private/tmp/err_sample
npm ERR! command failed
npm ERR! command sh -c textlint ./sample.tex
```

# Environment

```console
$  npm list
err_sample@ /private/tmp/err_sample
├── textlint-plugin-latex2e@1.0.4
├── textlint-rule-no-nfd@1.0.2
└── textlint@11.7.7
```

--- a/src/completeComment.ts
+++ b/src/completeComment.ts
@@ -20,7 +20,7 @@ import {
   AnyTxtNode,
   TxtNode,
   TxtParentNode,
-  TxtNodeLineLocation,
+  TextNodeRange,
 } from "@textlint/ast-node-types";
 
 export const convertCommentToTxtNode = (
@@ -56,35 +56,17 @@ export const convertCommentToTxtNode = (
 };
 
 export const isAppearedBeforeNode = (
-  nodeLocation: TxtNodeLineLocation,
-  commentLocation: TxtNodeLineLocation
+  nodeRange: TextNodeRange,
+  commentRange: TextNodeRange
 ): boolean => {
-  if (nodeLocation.start.line >= commentLocation.end.line) {
-    // \begin{document} % comment
-    // \end{document}
-    if (nodeLocation.start.line === commentLocation.end.line) {
-      return nodeLocation.start.column >= commentLocation.end.column;
-    }
-    return true;
-  }
-  return false;
+  return nodeRange[0] >= commentRange[1]; 
 };
 
 export const isIncludedByNode = (
-  nodeLocation: TxtNodeLineLocation,
-  commentLocation: TxtNodeLineLocation
+  nodeRange: TextNodeRange,
+  commentRange: TextNodeRange
 ): boolean => {
-  if (
-    nodeLocation.start.line <= commentLocation.start.line &&
-    nodeLocation.end.line >= commentLocation.end.line
-  ) {
-    // \begin{document}\end{document} % comment
-    if (nodeLocation.end.line === commentLocation.end.line) {
-      return nodeLocation.end.column > commentLocation.start.column;
-    }
-    return true;
-  }
-  return false;
+  return nodeRange[0] <= commentRange[0] && nodeRange[1] >= commentRange[1]
 };
 
 export const isParentNode = (node: any): node is TxtParentNode => {
@@ -105,12 +87,12 @@ export const insertComment = (
   }
   for (let i = 0; i < nodes.length; i++) {
     // If the comment is appeared before the node, insert it before the node.
-    if (isAppearedBeforeNode(nodes[i].loc, comment.loc)) {
+    if (isAppearedBeforeNode(nodes[i].range, comment.range)) {
       nodes.splice(i, 0, comment);
       return nodes;
     }
     // If the comment is included in the node, try to insert it recursively.
-    if (isIncludedByNode(nodes[i].loc, comment.loc)) {
+    if (isIncludedByNode(nodes[i].range, comment.range)) {
       if (isParentNode(nodes[i])) {
         nodes[i].children = insertComment(comment, nodes[i].children);
         return nodes;

--- a/test/completeComment.test.ts
+++ b/test/completeComment.test.ts
@@ -3,8 +3,7 @@ import "jest";
 import {
   ASTNodeTypes,
   TxtNodeLineLocation,
-  TextNodeRange,
-  TxtParentNode
+  TextNodeRange
 } from "@textlint/ast-node-types";
 import { latexParser } from "latex-utensils";
 import {
@@ -91,65 +90,65 @@ describe("Whether the comment is appeared before the node", () => {
     //
     // \begin{document}
     // \end{document}
-    const commentLoc = makeLocation(1, 0, 2, 0);
-    const nodeLoc = makeLocation(3, 0, 4, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeTruthy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeFalsy();
+    const commentRange = makeRange(0, 10);
+    const nodeRange = makeRange(11, 42);
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeTruthy();
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeFalsy();
   });
-  test("just before the node", async () => {
+  test("just before the node", async () => {	
     // % comment
-    // \begin{document}
-    // \end{document}
-    const commentLoc = makeLocation(1, 0, 2, 0);
-    const nodeLoc = makeLocation(2, 0, 3, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeTruthy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeFalsy();
+    // \begin{ document }	
+    // \end{document}	
+    const commentRange = makeRange(0, 10);	
+    const nodeRange = makeRange(10, 41);	
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeTruthy();	
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeFalsy();	
   });
   test("the comment after the node", async () => {
     // \begin{document}
     // \end{document}
     //
     // % comment
-    const commentLoc = makeLocation(4, 0, 5, 0);
-    const nodeLoc = makeLocation(1, 0, 2, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeFalsy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeFalsy();
+    const commentRange = makeRange(33, 43);
+    const nodeRange = makeRange(0, 31);
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeFalsy();
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeFalsy();
   });
   test("just after the node", async () => {
-    // \begin{document}
-    // \end{document}% comment
-    const commentLoc = makeLocation(2, 14, 3, 0);
-    const nodeLoc = makeLocation(1, 0, 2, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeFalsy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeFalsy();
-  });
-  test("the comment included by the node (with line break)", async () => {
-    // \begin{document}
-    //
-    // % comment
-    //
-    // \end{document}
-    const commentLoc = makeLocation(3, 0, 4, 0);
-    const nodeLoc = makeLocation(1, 0, 5, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeFalsy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeTruthy();
-  });
-  test("the comment included by the node (on the same line)", async () => {
-    // \begin{document}% comment
-    // \end{document}
-    const commentLoc = makeLocation(1, 0, 2, 0);
-    const nodeLoc = makeLocation(1, 0, 2, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeFalsy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeTruthy();
+    // \begin{document}	
+    // \end{document}% comment	
+    const commentRange = makeRange(31, 41);	
+    const nodeRange = makeRange(0, 31);	
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeFalsy();	
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeFalsy();	
   });
   test("the comment included by the node", async () => {
     // \begin{document}
+    //
     // % comment
+    //
     // \end{document}
-    const commentLoc = makeLocation(2, 0, 3, 0);
-    const nodeLoc = makeLocation(1, 0, 3, 14);
-    expect(isAppearedBeforeNode(nodeLoc, commentLoc)).toBeFalsy();
-    expect(isIncludedByNode(nodeLoc, commentLoc)).toBeTruthy();
+    const commentRange = makeRange(18, 28);
+    const nodeRange = makeRange(0, 43);
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeFalsy();
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeTruthy();
+  });
+  test("the comment included by the node (on the same line)", async () => {	
+    // \begin{document}% comment	
+    // \end{document}	
+    const commentRange = makeRange(16, 26);	
+    const nodeRange = makeRange(0, 40);	
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeFalsy();	
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeTruthy();
+  });
+  test("the comment included by the node", async () => {
+    // \begin{document}	
+    // % comment	
+    // \end{document}	
+    const commentRange = makeRange(17, 27);
+    const nodeRange = makeRange(0, 41);
+    expect(isAppearedBeforeNode(nodeRange, commentRange)).toBeFalsy();
+    expect(isIncludedByNode(nodeRange, commentRange)).toBeTruthy();
   });
 });
 
@@ -347,6 +346,42 @@ describe("Test insertComment", () => {
     };
     const expected = JSON.parse(JSON.stringify(nodes));
     expected.push(comment);
+
+    const actual = insertComment(comment, nodes);
+    expect(actual).toMatchObject(expected);
+    for (const node of actual) {
+      ASTTester.test(node);
+    }
+  });
+  test("comment after text without any white space", async () => {
+    // https://github.com/textlint/textlint-plugin-latex2e/issues/52
+    // A%B
+    // C
+    const comment = {
+      type: ASTNodeTypes.Comment,
+      range: makeRange(1, 4),
+      loc: makeLocation(1, 2, 2, 1),
+      raw: "%B\n",
+      value: "B"
+    };
+    const nodes = [
+      {
+        type: ASTNodeTypes.Str,
+        range: makeRange(0, 1),
+        loc: makeLocation(1, 1, 2, 2),
+        raw: "A",
+        value: "A",
+      },
+      {
+        type: ASTNodeTypes.Str,
+        range: makeRange(4, 5),
+        loc: makeLocation(2, 1, 2, 2),
+        raw: "C",
+        value: "C",
+      },
+    ];
+    const expected = JSON.parse(JSON.stringify(nodes));
+    expected.splice(1, 0, comment);
 
     const actual = insertComment(comment, nodes);
     expect(actual).toMatchObject(expected);

--- a/test/latex-to-ast.test.ts
+++ b/test/latex-to-ast.test.ts
@@ -153,6 +153,17 @@ describe("TxtNode AST", () => {
     expect(actual.children[0].children[2].type).toBe(ASTNodeTypes.Comment);
     expect(actual.children[0].children[4].type).toBe(ASTNodeTypes.Comment);
   });
+  test("issue52", () => {
+    const code = `A%B
+C`;
+    const actual = parse(code);
+    ASTTester.test(actual);
+    expect(actual.children.length).toBe(1);
+    expect(actual.children[0].type).toBe(ASTNodeTypes.Paragraph);
+    expect(actual.children[0].children[0].type).toBe(ASTNodeTypes.Str);
+    expect(actual.children[0].children[1].type).toBe(ASTNodeTypes.Comment);
+    expect(actual.children[0].children[2].type).toBe(ASTNodeTypes.Str);
+  });
 });
 
 describe("Fixing document", () => {

